shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.dart;
+
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.plugin.common.BinaryMessenger;
+
+/**
+ * Message conduit for 2-way communication between Android and Dart.
+ * <p>
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ * <p>
+ * See {@link BinaryMessenger}, which sends messages from Android to Dart
+ * See {@link PlatformMessageHandler}, which handles messages to Android from Dart
+ */
+class DartMessenger implements BinaryMessenger, PlatformMessageHandler {
+  private static final String TAG = "DartMessenger";
+
+  private final FlutterJNI flutterJNI;
+  private final Map<String, BinaryMessenger.BinaryMessageHandler> messageHandlers;
+  private final Map<Integer, BinaryMessenger.BinaryReply> mPendingReplies = new HashMap<>();
+  private int mNextReplyId = 1;
+
+  DartMessenger(@NonNull FlutterJNI flutterJNI) {
+    this.flutterJNI = flutterJNI;
+    this.messageHandlers = new HashMap<>();
+  }
+
+  @Override
+  public void setMessageHandler(String channel, BinaryMessenger.BinaryMessageHandler handler) {
+    if (handler == null) {
+      messageHandlers.remove(channel);
+    } else {
+      messageHandlers.put(channel, handler);
+    }
+  }
+
+  @Override
+  public void send(String channel, ByteBuffer message) {
+    send(channel, message, null);
+  }
+
+  @Override
+  public void send(String channel, ByteBuffer message, BinaryMessenger.BinaryReply callback) {
+    int replyId = 0;
+    if (callback != null) {
+      replyId = mNextReplyId++;
```

Comment:
Is this always being called from the main thread?

Reply:
Does it need to be?  Do you think we need to annotate all of DartMessenger with @UiThread?

Reply:
I was worried about concurrency problems with `mNextReplyId`. I'm not really familiar with the annotation. At a glance it looks like it fails in some strange edge cases but would still be potentially useful to declare on the API.

Reply:
Added a comment in BinaryMessenger that all such methods should be invoked on the same thread.